src/V3CfgBuilder.cpp

```diff
@@ -0,0 +1,260 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Control flow graph (CFG) builder
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2025 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Ast.h"
+#include "V3Cfg.h"
+#include "V3EmitV.h"
+
+#include <deque>
+#include <unordered_map>
+#include <unordered_set>
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+class CFGBuilder final : VNVisitorConst {
+    using EdgeKind = ControlFlowGraphEdge::Kind;
+
+    // STATE
+    // The graph being built, or nullptr if failed to build one
+    std::unique_ptr<ControlFlowGraph> m_cfgp{new ControlFlowGraph};
+    BasicBlock* m_currBBp = nullptr;  // Current basic block to add statements to
+    // Continuation block for given JumpBlock
+    std::unordered_map<AstJumpBlock*, BasicBlock*> m_jumpBlockContp;
+
+    // METHODS
+    BasicBlock& addBasicBlock() { return *new BasicBlock{m_cfgp.get(), false, false}; }
+
+    void addEdge(BasicBlock& src, BasicBlock& dst, EdgeKind kind) {
+        // Do not create duplicate edges
+        for (V3GraphEdge& edge : src.outEdges()) {
+            UASSERT(edge.top() != &dst, "Adding duplicate edge to ControlFlowGraph");
+        }
+        new ControlFlowGraphEdge{m_cfgp.get(), &src, &dst, kind};
+    }
+
+    void nonRepresentable() {
+        if (!m_cfgp) return;
+        m_cfgp.reset();
+    }
+
+    void simpleStatement(AstNode* nodep, bool representable = true) {
+        if (!m_cfgp) return;
+        // If non-representable, reset graph
+        if (!representable) {
+            m_cfgp.reset();
+            return;
+        }
+        // Just add to current block
+        m_currBBp->stmtps().emplace_back(nodep);
+    }
+
+    // VISITORS
+
+    // Base case, conservatively assume non-representable
+    void visit(AstNode* nodep) override {
+        if (!m_cfgp) return;
+        UINFO(9, "Unhandled AstNode type " << nodep->typeName());
+        m_cfgp.reset();
+    }
+
+    // Representable control flow statements
+    void visit(AstIf* nodep) override {
```

Comment:
```suggestion
    void visit(AstNodeIf* nodep) override {
```

Even if not handling Config, safer to presume at some day may be new flavors.